src/modules/power_profiles_daemon.cpp

```diff
@@ -0,0 +1,214 @@
+void PowerProfilesDaemon::getAllPropsCb(Glib::RefPtr<Gio::AsyncResult>& r) {
+  try {
+    auto _ = powerProfilesProxy_->call_finish(r);
+    // Power-profiles-daemon responded something, we can assume it's
+    // available, we can safely attach the activeProfile monitoring
+    // now.
+    connected_ = true;
+    powerProfileChangeSignal_ = powerProfilesProxy_->signal_properties_changed().connect(
+        sigc::mem_fun(*this, &PowerProfilesDaemon::profileChangedCb));
+    populateInitState();
+    dp.emit();
+  } catch (const std::exception& err) {
+    spdlog::error("Failed to query power-profiles-daemon via dbus: {}", err.what());
+  }
+}
+
+void PowerProfilesDaemon::populateInitState() {
+  // Retrieve current active profile
+  Glib::Variant<std::string> profileStr;
+  powerProfilesProxy_->get_cached_property(profileStr, "ActiveProfile");
+
+  // Retrieve profiles list, it's aa{sv}.
+  using ProfilesType = std::vector<std::map<Glib::ustring, Glib::Variant<std::string>>>;
+  Glib::Variant<ProfilesType> profilesVariant;
+  powerProfilesProxy_->get_cached_property(profilesVariant, "Profiles");
+  Glib::ustring name;
+  Glib::ustring driver;
+  Profile profile;
+  for (auto& variantDict : profilesVariant.get()) {
+    if (auto p = variantDict.find("Profile"); p != variantDict.end()) {
+      name = p->second.get();
+    }
+    if (auto d = variantDict.find("Driver"); d != variantDict.end()) {
+      driver = d->second.get();
+    }
+    profile = {name, driver};
+    availableProfiles_.push_back(profile);
+  }
+
+  // Find the index of the current activated mode (to toggle)
+  std::string str = profileStr.get();
+  switchToProfile(str);
+
+  update();
+}
+
+void PowerProfilesDaemon::profileChangedCb(
+    const Gio::DBus::Proxy::MapChangedProperties& changedProperties,
+    const std::vector<Glib::ustring>& invalidatedProperties) {
+  // We're likely connected if this callback gets triggered.
+  // But better be safe than sorry.
+  if (connected_) {
+    if (auto activeProfileVariant = changedProperties.find("ActiveProfile");
+        activeProfileVariant != changedProperties.end()) {
+      std::string activeProfile =
+          Glib::VariantBase::cast_dynamic<Glib::Variant<std::string>>(activeProfileVariant->second)
+              .get();
+      switchToProfile(activeProfile);
+      update();
+    }
+  }
+}
+
+// Look for the profile str in our internal profiles list. Using a
+// vector to store the profiles ain't the smartest move
+// complexity-wise, but it makes toggling between the mode easy. This
+// vector is 3 elements max, we'll be fine :P
+void PowerProfilesDaemon::switchToProfile(std::string const& str) {
+  auto pred = [str](Profile const& p) { return p.name == str; };
+  this->activeProfile_ = std::find_if(availableProfiles_.begin(), availableProfiles_.end(), pred);
+  if (activeProfile_ == availableProfiles_.end()) {
+    throw std::runtime_error("FATAL, can't find the active profile in the available profiles list");
```

Comment:
I'm not sure throwing exception is necessary here. It will be caught by Glibmm exception handler anyways, with no consequences.
You can just log the error and hide the module on the next update.